### PR TITLE
sander/upgradeToNg13

### PR DESCRIPTION
**Pull request by SanderElias**
- chore(monorepo): finished NX migrate to 13. WIP, fixes in next commits
- fix(ng-lib): fix build errors from nx migration
- refactor(monorepo): partway done upgradiing to ng 13, still problems with ESM/CommonJS
- chore(monorepo): nX update
- fix(monorepo): upto ESM modules, PW is working, ppt and sps still broken
- chore(scully): still working on esm
- chore(scully): still ESM, but sps comes back to life, yet to figure out why results are incomplete.
- fix(monorepo) SPS now working, Fixed XhrFactory bug by providing my own
- fix(scully-plugin-puppeteer): fixed exit problem with ppt renderer
- fix(monorepo): make tests work with ESM

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/scullyio/scully/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
With Angular 13, the Angular libs are only being distributed with ESM modules. 
To be able to use the platform-server, Scully needs to be able to support this. This PR is upgrading to Angular 13, and updated the whole mono-repo to use ESM instead of commonJS (with exception of a few tools and the cypress-test suite) 
The current Typescript version does make that extra hard because if not set up exactly right, it will translate all dynamic `import(xxx)` statements to `require(xxx)`

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

From now on, all plugins need to be ESM modules too. So plugin authors need to update their packages to cater for this change.
Also, if you are using a `.js` file for your proxy config, it needs to be renamed to `.cjs` 
There are no breaking changes for using Scully. Af far as I could check, there is no change there, and it still works with Angular 12+

### CODE REVIEW

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -110,44 +110,13 @@ jobs:
...
         run: |
-          node ./dist/libs/scully/src/scully --tds --project=scully-docs
+          node --experimental-specifier-resolution=node ./dist/libs/scully/src/scully --tds --project=scully-docs
       - name: Store docs-site
         uses: actions/upload-artifact@v2
         with:
           name: static-sites-docs
           path: dist/static/doc-sites
 
-  buildPWSample:
```
Comment (Jefiozie) [inline .github/workflows/workflow.yml:120]: should this be removed?

File: apps/sample-blog/src/app/app.component.ts [hunk 2]
```diff
@@ -1,20 +1,36 @@
...
+    im.init();
+    appRef.isStable.pipe(
+      tap((stable) => {
+        console.log(`
+        stable: ${stable}
+        AppComponent: ${this.currentState}
+        loc: ${this.loc.path(true)}
+        url: ${this.route.snapshot.pathFromRoot.map((x) => x.url).join('/')}
+        route: ${this.router.url},
+        ps: ${platformStrategy.constructor.name}
```
Comment (Jefiozie) [inline apps/sample-blog/src/app/app.component.ts:31]: Can we remove this now that we have resolved the route issue

File: apps/sample-blog/src/app/app.sps.module.ts [hunk 3]
```diff
@@ -1,23 +1,17 @@
...
-
 @NgModule({
-  imports: [
-    BrowserModule.withServerTransition({ appId: 'serverApp' }),
-    AppModule,
-    ScullyPlatformServerModule,
-  ],
-  providers: [],
+  imports: [BrowserModule.withServerTransition({ appId: 'serverApp' }), AppModule, ScullyPlatformServerModule],
+  // providers: [{ provide: XhrFactory, useClass: ScullyXhrFactory }],
```
Comment (Jefiozie) [inline apps/sample-blog/src/app/app.sps.module.ts:14]: Can be removed

File: libs/ng-lib/tsconfig.lib.prod.json [hunk 4]
```diff
@@ -4,6 +4,6 @@
     "declarationMap": false
   },
   "angularCompilerOptions": {
-    "enableIvy": true
```
Comment (Jefiozie) [inline libs/ng-lib/tsconfig.lib.prod.json:7]: just checking, this is removed as from v13 and up this is by default true?

File: libs/platform-server/package.json [hunk 5]
```diff
@@ -5,9 +5,8 @@
     "@angular/common": ">12.2.0",
     "@angular/core": ">12.2.0",
     "@angular/platform-server": ">12.2.0",
-    "@scullyio/ng-lib": "x.x.x"
+    "@scullyio/ng-lib": "x.x.x",
```
Comment (Jefiozie) [inline libs/platform-server/package.json:8]: Is this for any version? I think it is best to add a version here .

File: libs/platform-server/tsconfig.json [hunk 6]
```diff
@@ -17,9 +17,20 @@
...
+    "skipLibCheck": true,
+    "allowJs": true,
+    "lib": ["ES2020", "DOM"],
+    "types": ["node"],
+    "moduleResolution": "node",
+    "module": "esnext",
+    "target": "esnext"
   },
   "angularCompilerOptions": {
+    "enableIvy": true,
```
Comment (Jefiozie) [inline libs/platform-server/tsconfig.json:32]: Can this be removed?

File: libs/scully/src/lib/fileHanderPlugins/markdown.ts [hunk 7]
```diff
@@ -1,25 +1,27 @@
...
+// import 'prismjs/components/prism-bash';
+// import 'prismjs/components/prism-css';
+// import 'prismjs/components/prism-javascript';
+// import 'prismjs/components/prism-json';
+// import 'prismjs/components/prism-markup';
+// import 'prismjs/components/prism-markdown';
+// import 'prismjs/components/prism-typescript';
+// import 'prismjs/components/prism-jsx';
+// import 'prismjs/components/prism-tsx';
+// import 'prismjs/components/prism-docker';
```
Comment (Jefiozie) [inline libs/scully/src/lib/fileHanderPlugins/markdown.ts:20]: can we remove the commented lines?

File: libs/scully/src/lib/utils/captureMessage.ts [hunk 8]
```diff
@@ -1,30 +1,35 @@
-import * as sentry from '@sentry/node';
-import { readDotProperty } from './scullydot';
+// import * as sentry from '@sentry/node';
```
Comment (Jefiozie) [inline libs/scully/src/lib/utils/captureMessage.ts:1]: can we remove this?

File: libs/scully/src/lib/utils/platform-server/ps-worker.ts [hunk 9]
```diff
@@ -39,17 +41,8 @@ const writeToFs = findPlugin(WriteToStorage);
...
-    const pkg = join(__dirname, '../../../package.json');
-    // console.log(pkg)
-    version = jsonc.parse(readFileSync(pkg).toString()).version || '0.0.0';
-  } catch (e) {
-    // this is only for internals builds
-    // version = jsonc.parse(readFileSync(join(__dirname, '../../../package.json')).toString()).version || '0.0.0';
-  }
   const extraProviders: StaticProvider[] = [
-    { provide: APP_INITIALIZER, multi: true, useFactory: domContentLoadedFactory, deps: [DOCUMENT] },
+    // { provide: APP_INITIALIZER, multi: true, useFactory: scullyReadyEventFiredFactory, deps: [DOCUMENT] },
```
Comment (Jefiozie) [inline libs/scully/src/lib/utils/platform-server/ps-worker.ts:45]: can this lin be removed?

**Review by Jefiozie — COMMENTED** [review]
I can't test it locally (see slack) let have a chat about this.

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (SanderElias) [reply .github/workflows/workflow.yml:120]: yes

File: apps/sample-blog/src/app/app.component.ts (hunk 2 above)
Comment (SanderElias) [reply apps/sample-blog/src/app/app.component.ts:31]: not yet. also its in our internal kitchen-sink, not dev facing

File: apps/sample-blog/src/app/app.sps.module.ts (hunk 3 above)
Comment (SanderElias) [reply apps/sample-blog/src/app/app.sps.module.ts:14]: same deal

File: libs/ng-lib/tsconfig.lib.prod.json (hunk 4 above)
Comment (SanderElias) [reply libs/ng-lib/tsconfig.lib.prod.json:7]: Yes, done by the nx schematics. not a manual thing

File: libs/platform-server/package.json (hunk 5 above)
Comment (SanderElias) [reply libs/platform-server/package.json:8]: we should make an todo for this. For now its ok, but you are not wring. However, the version should really matter, as long as its there.

File: libs/platform-server/tsconfig.json (hunk 6 above)
Comment (SanderElias) [reply libs/platform-server/tsconfig.json:32]: probably

File: libs/scully/src/lib/fileHanderPlugins/markdown.ts (hunk 7 above)
Comment (SanderElias) [reply libs/scully/src/lib/fileHanderPlugins/markdown.ts:20]: yes, but this is a follow-up

File: libs/scully/src/lib/utils/platform-server/ps-worker.ts (hunk 9 above)
Comment (SanderElias) [reply libs/scully/src/lib/utils/platform-server/ps-worker.ts:45]: same

File: libs/scully/src/lib/utils/captureMessage.ts (hunk 8 above)
Comment (SanderElias) [reply libs/scully/src/lib/utils/captureMessage.ts:1]: not yet, we should add an toto here too.

**Review by Jefiozie — APPROVED** [review]
I tested this with wsl and everything worked. The issue on windows cmd is still present. Approving it so that we can ouch forward.

**Comment by SanderElias** [comment]
Closing this, as i'm going to move to a 'next' release from the 'next' branch

**Pull request closed without merging**